src/brsTypes/components/RoAssociativeArray.ts

```diff
@@ -93,19 +92,51 @@ export class RoAssociativeArray extends BrsComponent implements BrsValue, BrsIte
         // method with the desired name separately? That last bit would work but it's pretty gross.
         // That'd allow roArrays to have methods with the methods not accessible via `arr["count"]`.
         // Same with RoAssociativeArrays I guess.
-        let indexValue = this.modeCaseSensitive ? index.value : index.value.toLowerCase();
-        return this.elements.get(indexValue) || this.getMethod(index.value) || BrsInvalid.Instance;
+        return (
+            this.findElement(index.value, isCaseSensitive) ||
+            this.getMethod(index.value) ||
+            BrsInvalid.Instance
+        );
     }
 
-    set(index: BrsType, value: BrsType) {
+    set(index: BrsType, value: BrsType, isCaseSensitive = false) {
         if (index.kind !== ValueKind.String) {
             throw new Error("Associative array indexes must be strings");
         }
-        let indexValue = this.modeCaseSensitive ? index.value : index.value.toLowerCase();
+        // override old key with new one
+        let oldKey = this.findElementKey(index.value);
+        if (!this.modeCaseSensitive && oldKey) {
+            this.elements.delete(oldKey);
+        }
+
+        let indexValue = isCaseSensitive ? index.value : index.value.toLowerCase();
         this.elements.set(indexValue, value);
         return BrsInvalid.Instance;
     }
 
+    /** if AA is in insensitive mode, it means that we should do insensitive search of real key */
+    private findElementKey(elementKey: string, isCaseSensitiveFind = false) {
+        let useCaseSensitiveSearch = this.modeCaseSensitive && isCaseSensitiveFind;
+        elementKey = useCaseSensitiveSearch ? elementKey : elementKey.toLowerCase();
+        if (this.elements.has(elementKey)) {
+            return elementKey;
+        }
+
+        let realElementKey = undefined;
+        for (let key of this.elements.keys()) {
+            if ((useCaseSensitiveSearch ? key : key.toLowerCase()) === elementKey) {
+                realElementKey = key;
+                break;
+            }
+        }
```

Comment:
This seems less than ideal, as it'll cause an `O(n)` lookup for every access once `setCaseInsensitive` is called.  I'm curious if there's an alternative solution we can take that won't require us to iterate across every element every time.  Could we maintain a case-insensitive `Map` next to the case-sensitive one, and write to both for every `set()`? It'd of course require twice as much memory, but would maintain constant time lookups.

Reply:
Hey, I even forget about it, a great catch. 
So I've taken your idea of using a second map and used it for mapping lower case keys to case sensitive keys. so it will not consume so much memory, will not introduce additional computation and almost saves previous performance. 

but there is still be `O(n)` when we delete keys in case sensitive mode, cuz we can delete key that was used as alias in `KeysMap`